File: utils/rpbcopy.rs
use std::io::{Read, Write, BufWriter, BufReader};
use std::net::{SocketAddrV4, Ipv4Addr, TcpListener};
use std::process::{Command, Stdio, exit};

fn write_to_clipboard(output: &Vec<u8>) {
    let copy = Command::new("xclip")
        .arg("-selection")
        .arg("clipboard")
        .stdin(Stdio::piped())
        .spawn()
        .unwrap();
    let mut copy_stdin = copy.stdin.unwrap();
    let mut writer = BufWriter::new(&mut copy_stdin);
    writer.write_all(output).unwrap();
}

fn read_from_clipboard() -> Vec<u8> {
    let paste = Command::new("xclip")
        .arg("-selection")
        .arg("clipboard")
        .arg("-o")
        .arg("2>")
        .arg("/dev/null")
        .stdout(Stdio::piped())
        .spawn()
        .unwrap();
    let mut paste_stdout = paste.stdout.unwrap();
    let mut reader = BufReader::new(&mut paste_stdout);
    let mut buf = Vec::new();
    reader.read_to_end(&mut buf).unwrap();
    buf
}

fn setup_listener(port: u16) -> TcpListener {
    let loopback = Ipv4Addr::new(127, 0, 0, 1);
    let socket = SocketAddrV4::new(loopback, port);
    TcpListener::bind(socket).unwrap()
}

fn is_paste(buf: &Vec<u8>) -> bool {
    let paste_pattern = "!@#999$%^";

    if buf.len() != paste_pattern.len() {
        return false;
    }
    if buf.as_slice() != paste_pattern.as_bytes() {
        return false;
    }
    true
}

fn accept_copy_paste_loop(listener: &TcpListener) -> () {
    loop {
        let (mut tcp_stream, _addr) = listener.accept().unwrap();
        let mut reader = BufReader::new(&mut tcp_stream);
        let mut buf = Vec::new();
        reader.read_to_end(&mut buf).unwrap();
        if is_paste(&buf) {
            let cbd = read_from_clipboard();
            let mut writer = BufWriter::new(&mut tcp_stream);
            writer.write_all(&cbd).unwrap();
        } else {
            write_to_clipboard(&buf);
        }
    }
}

pub fn main() {
    if std::env::args().len() != 2 {
        println!("Usage: {} port", std::env::args().nth(0).unwrap());
        exit(1);
    }
    let port = std::env::args().nth(1).unwrap();
    let port: u16 = port.parse().unwrap();
    let listener = setup_listener(port);
    accept_copy_paste_loop(&listener);
}

File: utils/rpbcopy.zig
const std = @import("std");
const Stream = std.net.Stream;

const print = std.debug.print;

const IP = "127.0.0.1";
const PORT: u16 = 10001;
const BUFSIZE = 8192;

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const addr = try std.net.Address.resolveIp(IP, PORT);
    var listener = try addr.listen(.{});

    while (true) {
        if (listener.accept()) |conn| {
            defer conn.stream.close();
            try handle_connection(conn.stream, allocator);
        } else |_| {}
    }
}

pub fn handle_connection(stream: Stream, allocator: std.mem.Allocator) !void {
    var buf: [BUFSIZE]u8 = undefined;
    const size = try stream.read(&buf);
    if (is_paste(buf[0..])) {
        // we are pasting
        try read_from_clipboard(stream, allocator);
    } else {
        // we are copying
        try write_to_clipboard(stream, &buf, size, allocator);
    }
}

fn is_paste(buf: []const u8) bool {
    const paste_pattern = "!@#999$%^";
    const min_len = @min(paste_pattern.len, buf.len);

    if (std.mem.eql(u8, buf[0..min_len], paste_pattern)) {
        return true;
    }
    return false;
}

fn read_from_clipboard(stream: Stream, allocator: std.mem.Allocator) !void {
    const res = try std.process.Child.run(.{
        .allocator = allocator,
        .argv = &[_][]const u8{"pbpaste"},
    });
    defer allocator.free(res.stdout);
    defer allocator.free(res.stderr);

    try stream.writeAll(res.stdout);
    try std.posix.shutdown(stream.handle, std.posix.ShutdownHow.send);
}

fn write_to_clipboard(
    stream: Stream,
    buf: []u8,
    size0: usize,
    allocator: std.mem.Allocator,
) !void {
    const argv: []const []const u8 = &[_][]const u8{"pbcopy"};

    var child = std.process.Child.init(argv, allocator);
    child.stdin_behavior = .Pipe;
    try child.spawn();

    var size = size0;
    while (size > 0) {
        try child.stdin.?.writeAll(buf[0..size]);
        size = try stream.read(buf);
    }
    child.stdin.?.close();
    child.stdin = null;
    _ = try child.wait();
}
